src/audio/igo_nr/igo_nr.c

```diff
@@ -0,0 +1,521 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2021 intelliGo Corporation. All rights reserved.
+//
+// Author: Fu-Yun TSUO <fy.tsuo@intelli-go.com>
+
+#include <ipc/control.h>
+#include <ipc/stream.h>
+#include <ipc/topology.h>
+#include <user/igo_nr.h>
+#include <user/trace.h>
+#include <sof/common.h>
+#include <sof/debug/panic.h>
+#include <sof/drivers/ipc.h>
+#include <sof/lib/alloc.h>
+#include <sof/lib/memory.h>
+#include <sof/lib/uuid.h>
+#include <sof/list.h>
+#include <sof/platform.h>
+#include <sof/string.h>
+#include <sof/audio/buffer.h>
+#include <sof/audio/component.h>
+#include <sof/audio/pipeline.h>
+#include <sof/audio/igo_nr/igo_nr_comp.h>
+#include <sof/trace/trace.h>
+#include <sof/ut.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+
+static const struct comp_driver comp_igo_nr;
+
+uint8_t igo_heap[IGO_MEM_SIZE];
+
+/* 696ae2bc-2877-11eb-adc1-0242ac120002 */
+DECLARE_SOF_RT_UUID("igo-nr", igo_nr_uuid,  0x696ae2bc, 0x2877, 0x11eb, 0xad, 0xc1,
+		    0x02, 0x42, 0xac, 0x12, 0x00, 0x02);
+
+DECLARE_TR_CTX(igo_nr_tr, SOF_UUID(igo_nr_uuid), LOG_LEVEL_INFO);
+
+static void igo_nr_capture_s16(struct igo_nr_comp_data *cd,
+			       const struct audio_stream *source,
+			       struct audio_stream *sink,
+			       int32_t src_frames,
+			       int32_t sink_frames)
+{
+	int32_t in_nch = source->channels;
+	int32_t out_nch = sink->channels;
+	int32_t i;
+	int32_t idx_in = 0;
+	int32_t idx_out = 0;
+	int16_t *x;
+	int16_t *y;
+
+	for (i = 0; i < src_frames; i++) {
+		x = audio_stream_read_frag_s16(source, idx_in);
+		cd->in[cd->in_wpt] = (*x);
+
+		cd->in_wpt++;
+		idx_in += in_nch;
+	}
```

Comment:
memcpy ? IIRC I think there is a buffer API that does this and also takes in the wrap.

Reply:
this is more like deinterleaving data one by one. not pure memcpy and no wrap around.

Reply:
I feel like we should make this component expect 1 channel in, using something like the selector, at an absolute minimum we need a comment

Reply:
Even for more than 1 channel in, this component still selects the first channel to process.

Reply:
and just copy the rest? that seems confusing to users, i feel like we should only consume what we intend to process

Reply:
@cujomalainey my understanding is that even if the component process the first channel and bypass the other channels, it still has to consume all the channels. Consume/produce function are expected to take care of rpt/wpt update for audio buffer. It's impossible to consume the first channel only because data are interleaved.

Reply:
You can read source buffer with channels count stride and just update to consumed with the size that includes the omitted channel(s).

Reply:
@singalsu sorry I am a bit confusing. the code is doing exactly what you are saying. what is the difference?

Reply:
Sorry, I missed that while just looked the previous comments. Please ignore my comment!

Reply:
No problem!

Reply:
@fuyuntsuo it's probably worth a comment at the start of each of these function explaining the deinterleaving and why it's needed. That way everyone should be aligned.

Reply:
@lgirdwood agree. will put comments to ensure everyone is aligned.